nrf-hal-common/src/pwm.rs

```diff
@@ -0,0 +1,1000 @@
+        self.pwm.get_duty_on_group(self.group)
+    }
+    pub fn get_duty_off(&self) -> u16 {
+        self.pwm.get_duty_off_group(self.group)
+    }
+}
+
+impl<'a, T: sealed::Instance> embedded_hal::PwmPin for PwmGroup<'a, T> {
+    type Duty = u16;
+
+    fn disable(&mut self) {
+        Self::disable(self);
+    }
+
+    fn enable(&mut self) {
+        Self::enable(self);
+    }
+
+    fn get_duty(&self) -> Self::Duty {
+        self.get_duty_on()
+    }
+
+    fn get_max_duty(&self) -> Self::Duty {
+        self.get_max_duty()
+    }
+
+    fn set_duty(&mut self, duty: u16) {
+        self.set_duty_on(duty)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum Channel {
+    C0,
+    C1,
+    C2,
+    C3,
+}
+impl From<Channel> for isize {
+    fn from(variant: Channel) -> Self {
+        variant as _
+    }
+}
+impl From<Channel> for usize {
+    fn from(variant: Channel) -> Self {
+        variant as _
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum Group {
+    G0,
+    G1,
+}
+impl From<Group> for isize {
+    fn from(variant: Group) -> Self {
+        variant as _
+    }
+}
+impl From<Group> for usize {
+    fn from(variant: Group) -> Self {
+        variant as _
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum LoadMode {
+    Common,
+    Grouped,
+    Individual,
+    Waveform,
+}
+impl From<LoadMode> for u8 {
+    fn from(variant: LoadMode) -> Self {
+        variant as _
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum Seq {
+    Seq0,
+    Seq1,
+}
+impl From<Seq> for usize {
+    fn from(variant: Seq) -> Self {
+        variant as _
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum Prescaler {
+    Div1,
+    Div2,
+    Div4,
+    Div8,
+    Div16,
+    Div32,
+    Div64,
+    Div128,
+}
+impl From<Prescaler> for u8 {
+    fn from(variant: Prescaler) -> Self {
+        variant as _
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum CounterMode {
+    Up,
+    UpAndDown,
+}
+impl From<CounterMode> for bool {
+    fn from(variant: CounterMode) -> Self {
+        match variant {
+            CounterMode::Up => false,
+            CounterMode::UpAndDown => true,
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum Loop {
+    Disabled,
+    Times(u16),
+    Inf,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum PwmEvent {
+    Stopped,
+    LoopsDone,
+    PwmPeriodEnd,
+    SeqStarted(Seq),
+    SeqEnd(Seq),
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum StepMode {
+    Auto,
+    NextStep,
+}
+impl From<StepMode> for bool {
+    fn from(variant: StepMode) -> Self {
+        match variant {
+            StepMode::Auto => false,
+            StepMode::NextStep => true,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    DMABufferNotInDataMemory,
+}
+
+mod sealed {
+    use core::ops::Deref;
+    pub trait Instance: Deref<Target = crate::pac::pwm0::RegisterBlock> {}
```

Comment:
Making the trait private here makes it impossible to generically name the type  `Pwm<T>` in user code.

```rust
fn buzzer<T: pwm::sealed::Instance>(pwm: &mut pwm::Pwm<T>) {
```

complains that `sealed` is private.

Reply:
That is the idea of this design. You use a sealed module when its types should stay inaccessible :) I am not sure what the exact intentions are in this case tho :)

Reply:
This makes it impossible to write a function that "does something with a PWM instance". I would like to do this:

```rust
fn buzzer<T: pwm::sealed::Instance>(pwm: &mut pwm::Pwm<T>) { .. }
```

but with the current API, I'm forced to copy-paste the function 4 times.

```rust
fn buzzer_pwm0(pwm: &mut pwm::Pwm<PWM0>) { .. }
fn buzzer_pwm1(pwm: &mut pwm::Pwm<PWM1>) { .. }
fn buzzer_pwm2(pwm: &mut pwm::Pwm<PWM2>) { .. }
fn buzzer_pwm3(pwm: &mut pwm::Pwm<PWM3>) { .. }
```

which is not ideal.

If you want to seal the trait, a better alternative would be to keep it public, but have it require a private trait, like [this](https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed). 

This way client code can still name the trait, but not add new impls.

Reply:
Good idea! I was mostly just following the pattern used in the rest of this HAL, but i agree this is more useful